### PR TITLE
feat read file range

### PR DESCRIPTION
**Pull request by samhvw8**
## Context

There is many case that user want Roo read sample of file or read file in range of line

## Screenshots

| before | after |
| ------ | ----- |
|        |       |

## Get in Touch

----

> [!IMPORTANT]
> Adds `read_file_range` tool to read specific line ranges from files, with updated tool descriptions and tests.
> 
>   - **Behavior**:
>     - Adds `read_file_range` tool to `Cline` class in `Cline.ts` for reading specific line ranges from files.
>     - Updates `toolUseNames` and `toolParamNames` in `index.ts` to include `read_file_range`.
>     - Adds `getReadFileRangeDescription()` in `read-file-range.ts` for tool description.
>   - **Tests**:
>     - Adds tests for `readLines()` in `nthline.test.ts` to validate line range reading, including edge cases like negative and non-integer line numbers.
>   - **Misc**:
>     - Implements `readLines()` function in `nthline.ts` to read specified line ranges from files.
>     - Updates `TOOL_DISPLAY_NAMES` and `TOOL_GROUPS` in `tool-groups.ts` to include `read_file_range`.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=RooVetGit%2FRoo-Code&utm_source=github&utm_medium=referral)<sup> for 9099f87199323125e269c58b2289adde4e74a771. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

**Comment by mrubens** [comment]
Instead of adding a new tool and having the LLM choose between them, should we just add optional start/end line to the `read_file` tool along with a setting for the maximum number of lines to read from a file?

**Comment by samhvw8** [comment]
> Instead of adding a new tool and having the LLM choose between them, should we just add optional start/end line to the `read_file` tool along with a setting for the maximum number of lines to read from a file?

ok, i will update pr with that approach

**Comment by samhvw8** [comment]
@mrubens  updated ! :D

**Comment by mrubens** [comment]
Nice! How does it feel?

**Comment by samhvw8** [comment]
@mrubens current test with file 333M (2000000 line) from https://www.datablist.com/learn/csv/download-sample-csv-files, and it work ok, i think we can support mention read file in range with syntax like "@test.log::{10,100}"

image

**Comment by mrubens** [comment]
A couple thoughts on this:

1. I think we should have a default (that can be configured in settings) where if you don't specific the start and end it reads up to X lines of the file. I think that would save a lot of tokens from accidentally reading long files.
2. Along with that, I think we would need to return more information about the file along with the selected range of lines. Specifically how many total lines there are in the file, and ideally information from the tree-sitter about the definitions within the file and which lines they're on (similar to what Codey does).

What do you think?

**Comment by mrubens** [comment]
This is what I'm thinking of: https://github.com/ccrvlh/codey/blob/f823be88322b7417d75c4db98a0cc224de6ad97b/src/core/tools.ts#L306-L341

**Comment by samhvw8** [comment]
> A couple thoughts on this:
> 
> 1. I think we should have a default (that can be configured in settings) where if you don't specific the start and end it reads up to X lines of the file. I think that would save a lot of tokens from accidentally reading long files.
> 2. Along with that, I think we would need to return more information about the file along with the selected range of lines. Specifically how many total lines there are in the file, and ideally information from the tree-sitter about the definitions within the file and which lines they're on (similar to what Codey does).
> 
> What do you think?

1. sound good idea, let me do that
2. let me thinking how can we do it

**Comment by mrubens** [comment]
Great! Maybe we can reuse what that list_code_definitions tool does somehow?

**Comment by samhvw8** [comment]
> Great! Maybe we can reuse what that list_code_definitions tool does somehow?

yes, but the tool use processing is hard to maintain now, we need to separate into another module to easy maintain later

File: src/core/webview/ClineProvider.ts [hunk 1]
```diff
@@ -2272,6 +2277,7 @@ export class ClineProvider implements vscode.WebviewViewProvider {
 			telemetryKey,
 			machineId,
 			showRooIgnoredFiles: showRooIgnoredFiles ?? true,
+			maxReadFileLine: maxReadFileLine ?? 4500,
```
Comment (mrubens) [inline src/core/webview/ClineProvider.ts:2280]: I think the default should be 500

Comment (mrubens) [reply src/core/webview/ClineProvider.ts:2280]: Er, let's wait to make a call - but 4500 feels huge

File: src/services/tree-sitter/index.ts [hunk 2]
```diff
@@ -5,6 +5,59 @@ import { LanguageParser, loadRequiredLanguageParsers } from "./languageParser"
...
+	// C#
+	"cs",
+	// Ruby
+	"rb",
+	"java",
+	"php",
+	"swift",
+].map((e) => `.${e}`)
+
+export async function parseSourceCodeDefinitionsForFile(
```
Comment (mrubens) [inline src/services/tree-sitter/index.ts:32]: How can we change this to include the line numbers for each definition? Feels critical

File: src/core/Cline.ts [hunk 3]
```diff
@@ -2223,12 +2266,63 @@ export class Cline {
...
+										sourceCodeDef = `\n\n${result}`
+									}
+								} else {
+									// Read entire file
+									content = await extractTextFromFile(absolutePath)
+								}
+
+								// Add truncation notice if applicable
+								if (isFileTruncated) {
+									content += `\n\n[File truncated: showing ${maxReadFileLine} of ${totalLines} total lines]${sourceCodeDef}`
```
Comment (mrubens) [inline src/core/Cline.ts:2323]:
```suggestion
									content += `\n\n[File truncated: showing ${maxReadFileLine} of ${totalLines} total lines. Use start_line and end_line if you need to read more.].${sourceCodeDef}`
```

**Comment by mrubens** [comment]
I think for this to work well we're also going to want to change the search_files tool to include line numbers

File: src/services/tree-sitter/index.ts (hunk 2 above)
Comment (samhvw8) [reply src/services/tree-sitter/index.ts:32]: Let me try to do it 💪🏻

File: src/services/tree-sitter/index.ts [hunk 4]
```diff
@@ -106,6 +137,7 @@ async function parseFile(
 	filePath: string,
 	languageParsers: LanguageParser,
 	rooIgnoreController?: RooIgnoreController,
+	addLineNumbers?: boolean,
```
Comment (mrubens) [inline src/services/tree-sitter/index.ts:140]: Is there a reason to not just always add them?

File: src/services/tree-sitter/index.ts [hunk 5]
```diff
@@ -148,12 +180,20 @@ async function parseFile(
 
 			// Add separator if there's a gap between captures
 			if (lastLine !== -1 && startLine > lastLine + 1) {
-				formattedOutput += "|----\n"
+				if (addLineNumbers) {
+					formattedOutput += "||    ||----\n"
```
Comment (mrubens) [inline src/services/tree-sitter/index.ts:184]: Might be nice to add some logic around `maxLineNumberWidth` like https://github.com/RooVetGit/Roo-Code/blob/99f64cada15972cea60cf243b8a7704d8331fad7/src/integrations/misc/extract-text.ts#L57-L66

File: src/services/tree-sitter/index.ts [hunk 6]
```diff
@@ -148,12 +180,20 @@ async function parseFile(
...
+				} else {
+					formattedOutput += "|----\n"
+				}
 			}
 			// Only add the first line of the definition
 			// query captures includes the definition name and the definition implementation, but we only want the name (I found discrepencies in the naming structure for various languages, i.e. javascript names would be 'name' and typescript names would be 'name.definition)
 			if (name.includes("name") && lines[startLine]) {
-				formattedOutput += `│${lines[startLine]}\n`
+				if (addLineNumbers) {
+					formattedOutput += `│| ${startLine} ||${lines[startLine]}\n`
```
Comment (mrubens) [inline src/services/tree-sitter/index.ts:193]: I think this will be even better if you include the end line so it knows exactly what to read in

```suggestion
					formattedOutput += `│| ${startLine} - ${endLine} ||${lines[startLine]}\n`
```

Comment (mrubens) [reply src/services/tree-sitter/index.ts:193]: Er... this just adds the same line number twice, so that's not exactly right. Is there some way to get the end line of the method though?

File: src/services/tree-sitter/index.ts (hunk 4 above)
Comment (samhvw8) [reply src/services/tree-sitter/index.ts:140]: > Is there a reason to not just always add them?

no, i just don't know how it effect another function

Comment (samhvw8) [reply src/services/tree-sitter/index.ts:140]: let add start_line and end_line by default ?

File: src/core/prompts/__tests__/__snapshots__/system.test.ts.snap [hunk 7]
```diff
@@ -392,19 +392,47 @@ Always adhere to this format for the tool use to ensure proper parsing and execu
...
+Description: Request to read the contents of a file at the specified path. Use this when you need to examine the contents of an existing file you do not know the contents of, for example to analyze code, review text files, or extract information from configuration files. The output includes line numbers prefixed to each line (e.g. "1 | const x = 1"), making it easier to reference specific lines when creating diffs or discussing code. By specifying start_line and end_line parameters, you can efficiently read specific portions of large files without loading the entire file into memory. Automatically extracts raw text from PDF and DOCX files. May not be suitable for other types of binary files, as it returns the raw content as a string.
 Parameters:
 - path: (required) The path of the file to read (relative to the current working directory /test/path)
+- start_line: (optional) The starting line number to read from (1-based). If not provided, it starts from the beginning of the file.
+- end_line: (optional) The ending line number to read to (1-based, inclusive). If not provided, it reads to the end of the file.
```
Comment (mrubens) [inline src/core/prompts/__tests__/__snapshots__/system.test.ts.snap:399]: Does it read to the end of the file? Or is it limited by the max lines to read?

File: src/core/prompts/__tests__/__snapshots__/system.test.ts.snap [hunk 8]
```diff
@@ -392,19 +392,47 @@ Always adhere to this format for the tool use to ensure proper parsing and execu
...
 <read_file>
 <path>File path here</path>
+<start_line>Starting line number (optional)</start_line>
+<end_line>Ending line number (optional)</end_line>
 </read_file>
 
-Example: Requesting to read frontend-config.json
+Examples:
+
+1. Reading an entire file:
```
Comment (mrubens) [inline src/core/prompts/__tests__/__snapshots__/system.test.ts.snap:409]: I think we should mention the max number of lines here too

**Comment by KJ7LNW** [comment]
Does `tail` behavior work? for example:

Reading the lines >= 1000 of a log file:
```
<read_file>
<path>logs/application.log</path>
<start_line>1000</start_line>
</read_file>
```

**Comment by mrubens** [comment]
> Does `tail` behavior work? for example:
> 
> Reading the lines >= 1000 of a log file:
> 
> ```
> <read_file>
> <path>logs/application.log</path>
> <start_line>1000</start_line>
> </read_file>
> ```

I think so? But maybe we would actually want to support start_line=-100 for this (100 from the end)?

**Comment by KJ7LNW** [comment]
this approach would allow multiple ranges, slicing:

```
<range>123:456</range>
<range>777:</range>
<range>:1111</range>
```

also see my related comment here: https://github.com/RooVetGit/Roo-Code/pull/1374#issuecomment-2725720272

**Comment by samhvw8** [comment]
should we allow read multiple file but limit their range ?

**Comment by samhvw8** [comment]
> this approach would allow multiple ranges, slicing:
> 
> ```
> <range>123:456</range>
> <range>777:</range>       # 777 until the end 
> <range>:1111</range>     # beginning until 1111 (inclusive) 
> ```
> 
> also see my related comment here: [#1374 (comment)](https://github.com/RooVetGit/Roo-Code/pull/1374#issuecomment-2725720272)
> 
> > I think so? But maybe we would actually want to support start_line=-100 for this (100 from the end)?
> 
> I like that notation too, perhaps:
> 
> ```
> <range>:-100</range>
> ```
@KJ7LNW  sound good idea 

@mrubens  should we allow read multiple file but limit their range ?

File: src/core/prompts/__tests__/__snapshots__/system.test.ts.snap (hunk 7 above)
Comment (samhvw8) [reply src/core/prompts/__tests__/__snapshots__/system.test.ts.snap:399]: i will limited by the max lines to read, if not specify read range

**Comment by KJ7LNW** [comment]
> @mrubens should we allow read multiple file but limit their range ?

what do you mean by "allow read multiple file" ?

**Comment by samhvw8** [comment]
> > @mrubens should we allow read multiple file but limit their range ?
> 
> what do you mean by "allow read multiple file" ?

in single request and response allow user can read multiple file 

> > but limit their range
> 
> hard-coded limits are a bad idea. let the AI choose, or create a config option for limits
No we have setting, and already have that in this PR, you can check it

**Comment by mrubens** [comment]
I like the ideas of the support for tailing and multiple file reads, but I think those can be follow-ups - there's already a lot in this PR.

**Comment by KJ7LNW** [comment]
it looks like this commit serious picked up a huge number of changes that are not related to this feature. can you rebase on `main`?

**Comment by samhvw8** [comment]
> it looks like this PR picked up a huge number of changes that are not related to this feature. 
> 
> 
> 
> can you rebase on `main`?
> 
> 
> 
> be sure to backup your work first:
> 
> - git brach read-file-range-before-rebase
> 
> - git fetch origin # assuming roo is origin
> 
> - git rebase origin/main
 Already rebase on main, you can check the git commit

**Comment by KJ7LNW** [comment]
> > it looks like this PR picked up a huge number of changes that are not related to this feature.
> > can you rebase on `main`?
> > be sure to backup your work first:
> > 
> > * git brach read-file-range-before-rebase
> > * git fetch origin # assuming roo is origin
> > * git rebase origin/main
> 
> Already rebase on main, you can check the git commit

that looks much better thank you!

**Comment by mrubens** [comment]
Going to run this as my editor today and see how it goes!

**Comment by KJ7LNW** [comment]
> Going to run this as my editor today and see how it goes!

I curious how the AI utilizes the file ranges!

**Comment by mrubens** [comment]
@samhvw8 looking at the exports, I don't think it's currently appending the definitions to the tool result when the read is truncated.

**Comment by samhvw8** [comment]
> @samhvw8 looking at the exports of my chats, I don't think it's currently appending the definitions to the tool result when the read is truncated.


Let me find a way to truncate more 💪🏻

**Comment by mrubens** [comment]
> > @samhvw8 looking at the exports of my chats, I don't think it's currently appending the definitions to the tool result when the read is truncated.
> 
> Let me find a way to truncate more 💪🏻

Oh I think I was confused!

**Review by mrubens — APPROVED** [review]
